### DISCUSSION TITLE
How to collect back-end ```emit`` data on the front-end?

### DISCUSSION THREAD

**Discussion by yk0n9** [tauri-apps/tauri#7872] [Q&A]
my code:
```rust
#[tauri::command]
fn send_data(window: tauri::Window) {
    let vec: Vec<Vec<f64>>;
    vec.into_iter().for_each(|data| window.emit("data", data).unwrap());
}
```

I want to wait for the loop listener to finish collecting them, what should I do?
something like this
```javascript
let vec = [];
listen('data', (event) => {
  vec.push(event.payload);
});
```

**Comment by FabianLars** [comment]
hmm, the first thing that comes to mind is to send a special sequence at the end similar to null terminated strings used in some programming languages.

**Reply by yk0n9** [reply]
Thanks for your reply, the problem I have is that I want to send a very large piece of data like this
```rust
#[derive(Serialize, Deserialize, Default, Debug, Clone)]
struct Data {
    a: Vec<Cache>,
    b: Vec<Cache>,
    c: Vec<Cache>,
}

#[derive(Serialize, Deserialize, Default, Debug, Clone)]
struct Cache {
    a: Vec<f64>,
}

#[tauri::command]
fn send_data(window: tauri::Window) -> Data {
    Data {
        a: vec![Cache { a: vec![123.123456789; 100] }; 170000],
        b: vec![Cache { a: vec![123.123456789; 100] }; 170000],
        c: vec![Cache { a: vec![123.123456789; 100] }; 170000],
    }
}
```
This results in a blank WebView with the following prompt
WebView Error: 
```
There is a problem with this page. Please try to access it again later.
you still can:
· Open new tab
· Refresh this page
Error code: STATUS BREAKPOINT
```
However, if the floating point data in the structure is small (such as 1.0), it can be sent normally
```rust
#[tauri::command]
fn send_data(window: tauri::Window) -> Data {
    Data {
        a: vec![Cache { a: vec![1.0; 100] }; 170000],
        b: vec![Cache { a: vec![1.0; 100] }; 170000],
        c: vec![Cache { a: vec![1.0; 100] }; 170000],
    }
}
```
Data obtained from the ```javascript``` console

![ec38c19db567d364371bde3252e35fa](https://github.com/tauri-apps/tauri/assets/66518156/0a968489-aa66-4fd6-915b-5cbbe5edc07b)

May I ask what is the cause of this problem?

That’s why I want to use ```emit``` to send data

Can I get some suitable or better solutions?

**Reply by FabianLars** [reply]
Hmm, that may indeed be too much data for v1's IPC. What surprises me is that you're seeing an error. Normally it would just block the UI until it's done.

Soooo, the **potential** workarounds:
- Try the Tauri v2 alpha where we've reworked the IPC. Expect breaking changes between releases though and ofc it's not audited yet.
- Stick to events. They will be just as slow but hopefully won't error out.
- Try Tauri's custom protocols, which is also what v2's ipc is based on (but in v1 it blocks the main thread more than it will in v2): https://docs.rs/tauri/latest/tauri/struct.Builder.html#method.register_uri_scheme_protocol - real world example: https://github.com/JonasKruckenberg/pisano/blob/main/src-tauri/src/main.rs#L43
- Something i always try to do: Don't send that much data. I try to keep everything on the Rust side and send only what's absolutely required to the js side. I know that this doesn't work for everyone and every use case, still something worth mentioning imo.
